src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -220,13 +220,21 @@ private void CheckInvariants()
             solutionAttributes ??= _solutionAttributes;
             projectIds ??= ProjectIds;
             idToProjectStateMap ??= _projectIdToProjectStateMap;
-            options ??= Options.WithLanguages(GetRemoteSupportedProjectLanguages(idToProjectStateMap));
             analyzerReferences ??= AnalyzerReferences;
             projectIdToTrackerMap ??= _projectIdToTrackerMap;
             filePathToDocumentIdsMap ??= _filePathToDocumentIdsMap;
             dependencyGraph ??= _dependencyGraph;
             var newFrozenSourceGeneratedDocumentState = frozenSourceGeneratedDocument.HasValue ? frozenSourceGeneratedDocument.Value : _frozenSourceGeneratedDocumentState;
 
+            // PERF: Only invoke WithLanguages if we have different set of project IDs (AddProject/RemoveProject operation)
+            Debug.Assert(idToProjectStateMap.Count != _projectIdToProjectStateMap.Count ||
+                idToProjectStateMap.Keys.SetEquals(_projectIdToProjectStateMap.Keys));
+            Debug.Assert(idToProjectStateMap.Count != _projectIdToProjectStateMap.Count ||
+                Options == Options.WithLanguages(GetRemoteSupportedProjectLanguages(idToProjectStateMap)));
+            options ??= idToProjectStateMap.Count != _projectIdToProjectStateMap.Count
```

Comment:
❓ Do we have an enforced invariant today that we cannot add and remove a project in a single operation? If we add support for such an operation in the future, how do we ensure this subtle feature doesn't break?

Reply:
do the debug asserts not ensure this?

Reply:
Yep, and with the changes as per Jason's suggestion to use object reference equality, the assertions are much more clear and the code itself will never do the wrong thing.